session.go

```diff
@@ -0,0 +1,309 @@
+package wsep
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+)
+
+// Session represents a `screen` session.
+type Session struct {
+	// close receives nil when the session should be closed.
+	close chan struct{}
+	// closing receives nil when the session has begun closing.  The underlying
+	// process may still be exiting.  closing implies ready.
+	closing chan struct{}
+	// command is the original command used to spawn the session.
+	command *Command
+	// configFile is the location of the screen configuration file.
+	configFile string
+	// done receives nil when the session has completely shut down and the process
+	// has exited.  done implies closing and ready.
+	done chan struct{}
+	// error hold any error that occurred while starting the session.
+	error error
+	// execer is used to spawn the session and ready commands.
+	execer Execer
+	// id holds the id of the session for both creating and attaching.
+	id string
+	// options holds options for configuring the session.
+	options *Options
+	// ready receives nil when the session is ready to be attached.  error must be
+	// checked after receiving on this channel.
+	ready chan struct{}
```

Comment:
you have 3 different "state" channels: `closing`, `done`, and `ready`.  It found it relatively hard to trace the logic about dependencies and convince myself that everything gets closed at most once, and that everything that needs to close gets closed.

It might be clearer to model the lifecycle as a finite state machine with states like: `starting`, 'ready`, `closing`, `done`.  A sync.Cond + mutex is used to signal state changes for the routines that wait on them.  The addition of the mutex will also let you close the data race writing to `s.error` as this always accompanies a state change.